lottery/src/lib.rs

```diff
@@ -0,0 +1,342 @@
+            self.token_address.unwrap(),
+            FtAction::Transfer {
+                from: *from,
+                to: *to,
+                amount: amount_tokens,
+            },
+            0,
+        )
+        .await
+        .expect("Error in transfer");
+    }
+
+    /// Called by a player in order to participate in lottery
+    /// Requirements:
+    /// * Lottery must be on
+    /// * Contribution must be greater than zero
+    /// * The player cannot enter the lottery more than once
+    /// Arguments:
+    /// * `amount`: value or tokens to participate in the lottery
+    async fn enter(&mut self, amount: u128) {
+        if self.lottery_is_on() && !self.player_exist() && amount > 0 {
+            let player = Player {
+                player_id: msg::source(),
+                balance: amount,
+            };
+
+            if self.token_address.is_some() {
+                self.transfer_tokens(&msg::source(), &exec::program_id(), amount)
+                    .await;
+
+                self.lottery_balance += amount;
+                debug!("Add in Fungible Token: {}", amount);
+            }
+
+            let player_index = self.players.len() as u32;
+            self.players.insert(player_index, player);
+            msg::reply(Event::PlayerAdded(player_index), 0);
+        } else {
+            panic!(
+                "enter(): Lottery on: {}  player exist: {} amount: {}",
+                self.lottery_is_on(),
+                self.player_exist(),
+                amount
+            );
+        }
+    }
+
+    /// Removes player from lottery
+    /// Requirements:
+    /// * Lottery has started and lottery time has not expired
+    /// * Player must be on the player list
+    /// * Player can only remove himself
+    /// Arguments:
+    /// * `index`: lottery player index
+    async fn leave_lottery(&mut self, index: u32) {
+        if self.lottery_is_on() {
+            if let Some(player) = self.players.get(&index) {
+                if player.player_id == msg::source() {
+                    if self.token_address.is_some() {
+                        let balance = player.balance;
+                        self.transfer_tokens(&exec::program_id(), &msg::source(), balance)
+                            .await;
+                    } else {
+                        msg::send_bytes(player.player_id, b"LeaveLottery", player.balance);
+                    }
+
+                    self.players.remove(&index);
+                } else {
+                    panic!(
+                        "leave_lottery(): ActorId's does not match: player: {:?}  msg::source(): {:?}",
+                        player.player_id,
+                        msg::source()
+                    );
+                }
+            } else {
+                panic!("leave_lottery(): Player {} not found", index);
+            }
+        } else {
+            panic!("leave_lottery(): Lottery on: {}", self.lottery_is_on());
+        }
```

Comment:
Maybe it is worth to rewrite these if-else blocks a bit:
```suggestion
        if !self.lottery_is_on() {
            panic!("leave_lottery(): Lottery on: {}", self.lottery_is_on());
        }

        if let Some(player) = self.players.get(&index) {
            if player.player_id != msg::source() {
                panic!(
                    "leave_lottery(): ActorId's does not match: player: {:?}  msg::source(): {:?}",
                    player.player_id,
                    msg::source()
                );
            }

            if self.token_address.is_some() {
                let balance = player.balance;
                self.transfer_tokens(&exec::program_id(), &msg::source(), balance)
                    .await;
            } else {
                msg::send_bytes(player.player_id, b"LeaveLottery", player.balance);
            }

            self.players.remove(&index);
        } else {
            panic!("leave_lottery(): Player {} not found", index);
        }
```